### PR TITLE
added Alee problem

### PR DESCRIPTION
**Pull request by maranicolae**
added solution for Alee - Infoarena with detailed explanation on why the chosen algorithm fits this kind of problem

### CODE REVIEW

File: lab/skel_graph/traversal/problems/alee/alee.cpp [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+#include <bits/stdc++.h>
+
+class Task {
+	
+ public:
+	
+	void solve() {
+		read_input();
+		print_output(get_result());
+
```
Comment (neatudarius) [inline lab/skel_graph/traversal/problems/alee/alee.cpp:17]: Have you run clang-format?

File: lab/skel_graph/traversal/problems/alee/alee.cpp [hunk 2]
```diff
@@ -0,0 +1,152 @@
+// Copyright 2020 Nicolae Mara-Ioana <maraioana9967@gmail.com>
+
+// Alee - Drumuri minime in graf
```
Comment (neatudarius) [inline lab/skel_graph/traversal/problems/alee/alee.cpp:3]: ### ### 
```suggestion
// alee - Drumuri minime in graf
```

File: lab/skel_graph/traversal/problems/alee/alee.cpp [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+		read_input();
+		print_output(get_result());
+	
+	}
+
+ private:
+	int n;
+	int m;
+	int source;
+	std::vector<std::vector<char>> graph;
```
Comment (neatudarius) [inline lab/skel_graph/traversal/problems/alee/alee.cpp:24]: Please run clang-format to fix indentation problems

File: lab/skel_graph/traversal/problems/alee/alee.cpp [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+        // vectorul de distante pornind de la poarta de start
+		std::vector<std::vector<int>> minPath(n + 1, 
+            std::vector<int>(n + 1, INT_MAX));
+
+        // in coada se vor retine nodurile descoperite, dar inca nefinalizate
+		std::queue<std::pair<int, int>> nodeQueue;
+
+        // vectorul ce reprezinta posibilele pozitii 
+        // ale vecinilor unui element din matrice
+        std::vector<std::pair<int, int>> directions;
```
Comment (neatudarius) [inline lab/skel_graph/traversal/problems/alee/alee.cpp:66]: We can declare this as `const` and initialization with `{ {1, 0}, {0, 1}, {-1, 0}, {0, -1}}`

File: lab/skel_graph/traversal/problems/alee/alee.cpp [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+        // initial toate nodurile sunt nevizitate
+		std::vector<std::vector<bool>> visited(n + 1, 
+            std::vector<bool>(n + 1, false));
+
+        // vectorul de distante pornind de la poarta de start
+		std::vector<std::vector<int>> minPath(n + 1, 
+            std::vector<int>(n + 1, INT_MAX));
+
+        // in coada se vor retine nodurile descoperite, dar inca nefinalizate
+		std::queue<std::pair<int, int>> nodeQueue;
```
Comment (neatudarius) [inline lab/skel_graph/traversal/problems/alee/alee.cpp:62]: Please rename this to `q` to be consistent with the theory example.

Same for `minPath` -> `d`.

File: lab/skel_graph/traversal/problems/alee/alee.cpp [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+                        // doar daca reprezinta o zona libera
+                        // un copac ar impiedica inaintarea pe un drum
+                        if (graph[x][y] == ' ') {
+                            nodeQueue.push({x, y});
+                        }
+                    }
+                }
+			}
+            // se elimina din coada
+			nodeQueue.pop();
```
Comment (neatudarius) [inline lab/skel_graph/traversal/problems/alee/alee.cpp:102]: This should be moved right after line 79.

File: lab/skel_graph/traversal/problems/alee/alee.cpp [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+	}
+};
+	
+int main() {
+    // din cauza ca fac redirectari, salvez starea lui cin si cout
+    auto cin_buff = std::cin.rdbuf();
+    auto cout_buff = std::cout.rdbuf();
+
+    // las liniile urmatoare daca citesc din fisier
+    std::ifstream fin("custi.in");
```
Comment (neatudarius) [inline lab/skel_graph/traversal/problems/alee/alee.cpp:133]: typo?

File: lab/skel_graph/traversal/problems/alee/alee.cpp [hunk 8]
```diff
@@ -0,0 +1,152 @@
...
+    std::cin.rdbuf(fin.rdbuf()); // save and redirect
+
+    // las liniile urmatoare daca afisez in fisier
+    std::ofstream fout("custi.out");
```
Comment (neatudarius) [inline lab/skel_graph/traversal/problems/alee/alee.cpp:137]: typo?

File: lab/skel_graph/traversal/problems/alee/main.cpp [hunk 9]
```diff
@@ -0,0 +1,157 @@
+// Copyright 2020 Nicolae Mara-Ioana <maraioana9967@gmail.com>
```
Comment (neatudarius) [inline lab/skel_graph/traversal/problems/alee/main.cpp:1]: Please check that you have a duplicate source code. Delete alee.cpp and if the reported issues are still available, fix them in main.cpp

**Review by neatudarius — CHANGES_REQUESTED** [review]

File: lab/skel_graph/traversal/problems/alee/Makefile [hunk 10]
```diff
@@ -0,0 +1,18 @@
...
+
+build:
+	$(CC) $(CFLAGS) $(LDFLAGS) $(SRC) -o $(EXE) 
+
+run: build
+	time ./$(EXE) && echo "pwp" &> /dev/null
+
+.PHONY: clean
+clean:
+	rm -f *.o *.exe *.out
```
Comment (neatudarius) [inline lab/skel_graph/traversal/problems/alee/Makefile:18]:
```suggestion
	rm -f *.o *.exe *.out

```

File: lab/skel_graph/traversal/problems/alee/_example.in [hunk 11]
```diff
@@ -0,0 +1,14 @@
...
+coordonatele la care se afla cei m copaci
+2 7
+3 3
+4 6
+5 4
+7 3
+7 5
+
+coordonatele portii de start si finish
+1 1 8 8
```
Comment (neatudarius) [inline lab/skel_graph/traversal/problems/alee/_example.in:14]:
```suggestion
1 1 8 8

```

**Review by neatudarius — CHANGES_REQUESTED** [review]

File: lab/skel_graph/traversal/problems/alee/alee.cpp [hunk 12]
```diff
@@ -0,0 +1,152 @@
...
+    std::pair<int, int> startGate;
+    std::pair<int, int> endGate;
+	
+    // s-a ales urmatoarea codificare:
+    // graph[i][j] = 'T', atunci exista un copcac la coordonatele i, j
+    // graph[i][j] = ' ', atunci exista zona libera la coordonatele i, j
+	void read_input() {	
+        std::cin >> n >> m;
+        // initial se marcheaza toate elementele din matrice ca fiind zone libere
+        graph = std::vector<std::vector<char>>(n + 1,
```
Comment (RaduNichita) [inline lab/skel_graph/traversal/problems/alee/alee.cpp:34]: I think we should keep the convention and replace graph with adj. @neatudarius what do you think?

File: lab/skel_graph/traversal/problems/alee/alee.cpp [hunk 13]
```diff
@@ -0,0 +1,152 @@
...
+		minPath[startGate.first][startGate.second] = 1;
+		nodeQueue.push({startGate.first, startGate.second});
+
+		while (nodeQueue.size()) {
+            // se extrage primul element din coada
+			auto node = nodeQueue.front();
+
+            // se parcurg vecinii acestuia folosind vectorul de directii
+            // acestia pot fi maximum patru
+			for (auto& dir : directions) {
```
Comment (RaduNichita) [inline lab/skel_graph/traversal/problems/alee/alee.cpp:83]: Can you use c++17 and structured bindings, please? That means for (auto &[dx, dy]: directions ?

File: lab/skel_graph/traversal/problems/alee/alee.cpp [hunk 14]
```diff
@@ -0,0 +1,152 @@
...
+        }
+
+        // returneaza distanta de la poarta de start la poarta destinatie
+		return minPath[endGate.first][endGate.second];
+	}
+	
+    // verifica daca coordonatele unui vecin sunt valide 
+    // adica nu depasesc extremele matricei
+    bool correctBounds(const std::pair<int, int>& p1, const std::pair<int, int>& p2) {
+        if (p1.first + p2.first <= n && p1.first + p2.first >= 1
```
Comment (RaduNichita) [inline lab/skel_graph/traversal/problems/alee/alee.cpp:112]: Maybe here we can directly return the result? I mean return (condition1 && condition2 && condition3 && condition4) ?

File: lab/skel_graph/traversal/problems/alee/alee.cpp [hunk 15]
```diff
@@ -0,0 +1,152 @@
...
+    delete task;
+
+    // restore pentru cin si cout
+    std::cin.rdbuf(cin_buff);
+    std::cout.rdbuf(cout_buff);
+
+    // obs. nu e nevoie sa inchid fisierele
+    // cand se apeleaza destructorii pentru fin si fout se vor inchide1
+    return 0;
+}
```
Comment (RaduNichita) [inline lab/skel_graph/traversal/problems/alee/alee.cpp:152]:
```suggestion

}
```

File: lab/skel_graph/traversal/problems/alee/alee.in [hunk 16]
```diff
@@ -0,0 +1,8 @@
+8 6
+2 7
+3 3
+4 6
+5 4
+7 3
+7 5
+1 1 8 8
```
Comment (RaduNichita) [inline lab/skel_graph/traversal/problems/alee/alee.in:8]:
```suggestion
1 1 8 8

```

File: lab/skel_graph/traversal/problems/alee/alee.cpp [hunk 17]
```diff
@@ -0,0 +1,152 @@
+// Copyright 2020 Nicolae Mara-Ioana <maraioana9967@gmail.com>
```
Comment (RaduNichita) [inline lab/skel_graph/traversal/problems/alee/alee.cpp:1]: Please remove this file and apply the suggestions made in main.cpp, as @neatudarius you have duplicate file

File: lab/skel_graph/traversal/problems/alee/alee.cpp [hunk 18]
```diff
@@ -0,0 +1,152 @@
...
+ public:
+	
+	void solve() {
+		read_input();
+		print_output(get_result());
+	
+	}
+
+ private:
+	int n;
```
Comment (RaduNichita) [inline lab/skel_graph/traversal/problems/alee/alee.cpp:21]: I think we can have int n, m,  source;

File: lab/skel_graph/traversal/problems/alee/alee.cpp [hunk 19]
```diff
@@ -0,0 +1,152 @@
...
+	void read_input() {	
+        std::cin >> n >> m;
+        // initial se marcheaza toate elementele din matrice ca fiind zone libere
+        graph = std::vector<std::vector<char>>(n + 1, 
+            std::vector<char>(n + 1, ' '));
+		for (int i = 1, x, y; i <= m; i++) {
+            // se citesc coordonatele la care se afla copacii
+            // se actualizeaza valorile din celulele in care se regasesc obstacole
+			std::cin >> x >> y;
+			graph[x][y] = 'T';
```
Comment (RaduNichita) [inline lab/skel_graph/traversal/problems/alee/alee.cpp:40]: I think it's may be a bit complicated to have 'T' and ' '. Maybe change to boolean and have 1 for tree and 0 for no tree?

**Pull request merged**